measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
 
-        p_cons = [cost < budget]
+        # warmstart function initializes the model with all the binary decisions values stored in the model
+        if self.obj == ObjectiveLib.A:
+            # use warmstart but without initializing grey-box block
+            grey_box_opt = False
+        else:
+            # use warmstart with initializing grey-box block
+            grey_box_opt = True
+
+        # use warmstart with grey-box block
+        self.customized_warmstart(grey_box=grey_box_opt)
 
-        #for idx in :
-        #    for t in :
-        #        p_cons += [j<=1]
+    def extract_store_sol(self, budget_opt, store_name):
+        """
+        Extract the solution from pyomo model, store in two pickles.
+        First data file (store_name+"_fim_"+budget_opt): pickle, a Nm*Nm numpy array that is the solution of measurements
+        Second data file (store_name+"_fim_"+budget_opt): pickle, a Np*Np numpy array that is the FIM
+
+        Arguments
+        ---------
+        :param store_name: if not None, store the solution and FIM in pickle file with the given name
+        """
+        fim_result = np.zeros(
+            (self.sens_info.n_parameters, self.sens_info.n_parameters)
+        )
+        for i in range(self.sens_info.n_parameters):
+            for j in range(i, self.sens_info.n_parameters):
+                fim_result[i, j] = fim_result[j, i] = pyo.value(
+                    self.mod.total_fim[i, j]
+                )
 
-        p_cons += []
+        print(fim_result)
+        print("trace:", np.trace(fim_result))
+        print("det:", np.linalg.det(fim_result))
+        print(np.linalg.eigvals(fim_result))
 
-        obj = cp.Maximize(objective())
+        ans_y, _ = self.extract_solutions()
+        print("pyomo calculated cost:", pyo.value(self.mod.cost))
+        print("if install dynamic measurements:")
+        print(pyo.value(self.mod.if_install_dynamic[3]))
+
+        if store_name:
+
+            file = open(store_name + str(budget_opt), "wb")
+            pickle.dump(ans_y, file)
+            file.close()
+
+            file2 = open(store_name + "_fim_" + str(budget_opt), "wb")
+            pickle.dump(fim_result, file2)
+            file2.close()
+
+    def _locate_initial_file(self, budget):
+        """
+        Given the budget, select which initial solution it should use by locating the closest budget that has stored solutions
+        It selects according to:
+        1) if this budget has a previous solution, find it, return the file name
+        2) if this budget does not have a previous solution, find the closest budget that has a solution, return the file name
+
+        Arguments
+        --------
+        budget: budget
+
+        Return
+        ------
+        y_init_file: the file name that contains a previous solution
+        """
+        ## find if there has been a original solution for the current budget
+        if budget in self.curr_res_list:  # use an existed initial solutioon
```

Comment:
Can all of this code get replaced by a scipy function? https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html

Reply:
yes. used KDTree now